fblckgen.1
.\" 
.\" Copyright (c) 2006 Paul Ripke. All rights reserved.
.\"
.\" This software is distributed under the so-called ``revised Berkeley
.\" License'':
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" This software is provided ``as is'', and any express or implied
.\" warranties, including, but not limited to, the implied warranties of
.\" merchantability and fitness for a particular purpose are disclaimed.
.\" In no event shall the author be liable for any direct, indirect,
.\" incidental, special, exemplary, or consequential damages (including,
.\" but not limited to, procurement of substitute goods or services;
.\" loss of use, data, or profits; or business interruption) however
.\" caused and on any theory of liability, whether in contract, strict
.\" liability, or tort (including negligence or otherwise) arising in
.\" any way out of the use of this software, even if advised of the
.\" possibility of such damage.
.\"
.\"
.TH FBLCKGEN 1 2.2 iotools
.SH NAME
.B fblckgen
\- Fast block generator
.SH SYNOPSIS
.B fblckgen
.RB [ \-a | \-r ]
.RB [ -q ]
.RB [ -v ]
.RB [ \-b
.IR blocksize ]
.RB [ \-c
.IR count ]
.SH DESCRIPTION
.B fblckgen
generates blocks of data, either a
repeating ascii sequence which is very compressible, or a pseudo-
random binary sequence, which, although very simple, does not
compress. Very handy for benchmarking tape drives, or just making a
sized lump of data. By using double buffering and either pthreads
or multiple processes, it can generally keep a tape drive busy.
.PP
Data is written directly to a copy of the standard output file descriptor. Shell
redirection must be used to output to an alternate destination.
.PP
Before exiting, 
.B fblckgen
writes a single summary line to standard error, containing statistics from the run.
.PP
.SH OPTIONS
.TP
.B \-a
Instructs
.B fblckgen
to generate blocks of a repeating ASCII sequence, from octal 040 (` ', space),
through octal 176 (`~', tilde). The sequence repeats, without newlines. This the
default mode of operation.
.TP
.BI \-b\  blocksize
Writes blocks of size
.I blocksize
to standard output. If not specified, a blocksize of `1s', 512 bytes, is used.
.TP
.BI \-c\  count
Writes
.I count
blocks, or until error if
.I count
is zero. If not specified, a default of `1k', 1024, is used.
.TP
.B \-q
Quiet operation, suppresses the printing of the summary line.
.TP
.B \-r
Instructs
.B fblckgen
to generate a pseudo-random sequence, based on an inlined copy of
.BR rand (3).
.\" x[i+1] = x[i] * 1103515245 + 12345
.TP
.B \-v
Verbose: regularly prints a status line showing current progress.
.LP
All numeric arguments may take an optional letter suffix, similar to the
.BR strsuftollx (3)
function introduced in NetBSD.
.RS
.PD 0
.TP
s
sector; multiply by 512.
.TP
k
kibibytes; multiply by 1024 or 2^10.
.TP
m
mebibytes; multiply by 1048576 or 2^20.
.TP
g
gibibytes; multiply by 2^30.
.TP
t
tebibytes; multiply by 2^40.
.TP 
p
pebibytes; multiply by 2^50.
.TP 
e
exbibytes; multiply by 2^60.
.PD    
.RE
.SH EXIT STATUS
.B fblckgen
exits 0 on success, and >0 if an error occurred.
.SH EXAMPLES
Filling an LTO1 tape with pseudo-random data:
.sp
.RS
.nf
sh$ fblckgen -rb 64k -c 0 > /dev/nrst1
Write failed: Input/output error
105722740736 bytes written in 7064.506 secs (14614.590 KB/sec)
.fi
.RE
.sp
.SH SEE ALSO
.BR iohammer (1),\  mbdd (1)
.SH AUTHOR
.B fblckgen
and the rest of the iotools were developed by Paul Ripke,
<stixpjr@gmail.com>. See <http://stix.id.au/wiki/iotools> for the current
version.


CHANGES
/***********
 * CHANGES *
 ***********/

2.2 released 2013-03-13
- minor fixes, silenced compiler warnings, cleaned up decaying average code.
- changed email address.

2.1 released 2008-10-20
- added the verbose status/progress bar to fblckgen(1), iohammer(1) and
  mbdd(1).
- added multiple destination support for mbdd(1). Think high-speed
  tee(1).
- added a block copy limit to mbdd(1).
- fixed a bunch of bugs.

2.0 released 2006-07-26
- migrated to GNU autoconf.
- added the mbdd(1) tool (pthread only).
- added man pages.
- tested on a variety of platforms.

1.0 largely unreleased 2003-10-12
- contained only fblckgen(1) and iohammer(1).
- used fork(2) style multi-processing or pthreads.


fblckgen.c
/*
 * Copyright (c) 2004 Paul Ripke. All rights reserved.
 *
 *  This software is distributed under the so-called ``revised Berkeley
 *  License'':
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * This software is provided ``as is'', and any express or implied
 * warranties, including, but not limited to, the implied warranties of
 * merchantability and fitness for a particular purpose are disclaimed.
 * In no event shall the author be liable for any direct, indirect,
 * incidental, special, exemplary, or consequential damages (including,
 * but not limited to, procurement of substitute goods or services;
 * loss of use, data, or profits; or business interruption) however
 * caused and on any theory of liability, whether in contract, strict
 * liability, or tort (including negligence or otherwise) arising in
 * any way out of the use of this software, even if advised of the
 * possibility of such damage.
 */

#include "iotools.h"
#include "common.h"

/* Prototypes */
static void	*makeBlocks(void *);
static void	*status(void *);
static void	cleanup(int);
static void	usage();

/* Globals */
static char *buf[2];
static int type;
static int flAborted;
static int flFinished;
static long blockSize;
static int64_t numBlocks;
static int64_t blocksWritten;
static int outfd;

#ifdef USE_PTHREADS
static pthread_mutex_t lock;
static pthread_cond_t less;
static pthread_cond_t more;
static int nblocks;
#else
static int ctl1[2], ctl2[2];
#endif

int
main(int argc, char **argv)
{
	int c;
	int flQuiet;
	int flVerbose;
	int64_t numWritten;
	double duration;
	void *buffer;
	struct timeval tpstart, tpend;
#ifdef USE_PTHREADS
	pthread_t tid;
	pthread_attr_t attr;
	pthread_t status_tid;
#else
	pid_t pid;
#endif
	/* close off stdout, so stdio doesn't play with it */
	outfd = dup(STDOUT_FILENO);
	if (outfd < 0) {
		perror("dup on stdout failed");
		exit(1);
	}
	close(STDOUT_FILENO);

	/* set defaults */
	blockSize = 512;
	numBlocks = 1024;
	type = ALPHADATA;
	flQuiet = 0;
	flVerbose = 0;

	while ((c = getopt(argc, argv, "arb:c:qv")) != EOF) {
		switch (c) {
		case 'a':
			type = ALPHADATA;
			break;
		case 'r':
			type = RANDDATA;
			break;
		case 'b':
			blockSize = getnum(optarg);
			break;
		case 'c':
			numBlocks = getnum(optarg);
			break;
		case 'q':
			flQuiet = 1;
			break;
		case 'v':
			flVerbose = 1;
			break;
		case '?':
		default:
			usage();
			exit(1);
		}
	}

	/* Double buffering, get twice the block size of memory */
#ifdef USE_PTHREADS
	/* Just use a normal malloc for threads */
	if ((buffer = malloc(blockSize * 2)) == NULL) {
		fprintf(stderr, "malloc for %ld bytes failed.\n",
		    blockSize * 2);
		exit(1);
	}
#else
	/* Shared memory for processes */
	buffer = getshm(blockSize * 2);
#endif
	buf[0] = buffer;
	buf[1] = (char *)buffer + blockSize;
	flAborted = 0;
#ifdef USE_PTHREADS
	nblocks = 0;
	MYASSERT(pthread_mutex_init(&lock, NULL) == 0,
	    "pthread_mutex_init failed");
	MYASSERT(pthread_cond_init(&less, NULL) == 0,
	    "pthread_cond_init failed");
	MYASSERT(pthread_cond_init(&more, NULL) == 0,
	    "pthread_cond_init failed");
	MYASSERT(pthread_attr_init(&attr) == 0,
	    "pthread_attr_init failed");
	MYASSERT(pthread_attr_setdetachstate(&attr,
	    PTHREAD_CREATE_DETACHED) == 0,
	    "pthread_attr_setdetachstate failed");
	MYASSERT(pthread_create(&tid, &attr, &makeBlocks, NULL) == 0,
	    "pthread_create failed");
	MYASSERT(pthread_attr_destroy(&attr) == 0,
	    "pthread_attr_destroy failed");
	if (flVerbose) {
		MYASSERT(pthread_create(&status_tid, NULL, &status, NULL) == 0,
		    "pthread_create failed");
	}
#else
	MYASSERT(pipe(&ctl1[0]) == 0, "pipe failed");
	MYASSERT(pipe(&ctl2[0]) == 0, "pipe failed");
	if ((pid = fork()) == -1) {
		perror("fork failed");
		_exit(1);
	}
	if (pid == 0) {
		signal(SIGINT, SIG_IGN);
		makeBlocks(NULL);
		_exit(0);
	}
#endif
	signal(SIGINT, &cleanup);
	MYASSERT(gettimeofday(&tpstart, NULL) == 0, "gettimeofday failed");
	for (blocksWritten = 0; !flAborted &&
	    (numBlocks == 0 || blocksWritten < numBlocks); blocksWritten++) {
		/* tell child to make another */
		/* and wait for block of data */
#ifdef USE_PTHREADS
		MYASSERT(pthread_mutex_lock(&lock) == 0,
		    "pthread_mutex_lock failed");
		while (nblocks == 0)
			MYASSERT(pthread_cond_wait(&more, &lock) == 0,
			    "pthread_cond_wait failed");
		MYASSERT(pthread_mutex_unlock(&lock) == 0,
		    "pthread_mutex_unlock failed");
#else
		MYASSERT(read(ctl1[0], &c, 1) == 1, "read on pipe failed");

		MYASSERT(write(ctl2[1], &c, 1) == 1, "Write on pipe failed");
#endif
		/* write it */
		numWritten = write(outfd, buf[blocksWritten & 1], blockSize);
		if (numWritten != blockSize) {
			if (numWritten > 0) {
				fprintf(stderr, "Short write: "
				    "%" PRId64 " bytes\n", (int64_t)numWritten);
			} else
				perror("Write failed");
			break;
		}
#ifndef USE_PTHREADS
		if (flVerbose)
			statusLine(blocksWritten * blockSize / 1024,
				numBlocks * blockSize / 1024,
				"KiB", "KiB/s");
#endif
#ifdef USE_PTHREADS
		MYASSERT(pthread_mutex_lock(&lock) == 0,
		    "pthread_mutex_lock failed");
		nblocks--;
		MYASSERT(pthread_cond_signal(&less) == 0,
		    "pthread_cond_signal failed");
		MYASSERT(pthread_mutex_unlock(&lock) == 0,
		    "pthread_mutex_unlock failed");
#endif
		
	}
	MYASSERT(gettimeofday(&tpend, NULL) == 0, "gettimeofday failed");
	flFinished = 1;
#ifdef USE_PTHREADS
	if (flVerbose)
		pthread_join(status_tid, NULL);
#endif
	if (flAborted)
		fprintf(stderr, "Transfer aborted.\n");
	if (!flQuiet) {
		duration = tpend.tv_sec - tpstart.tv_sec +
		    (tpend.tv_usec - tpstart.tv_usec) / 1000000.0;
		fprintf(stderr, "%" PRId64
		    " bytes written in %.3lf secs (%.3lf KiB/sec)\n",
		    (int64_t)blocksWritten * blockSize,
		    duration, (double)blocksWritten * blockSize / duration / 1024.0);
	}
	if (flAborted)
#ifdef USE_PTHREADS
		pthread_cond_signal(&less);
#else
		kill(pid, SIGTERM);
#endif

	return 0;
}

static void *
makeBlocks(void *dummy)
{
#ifndef USE_PTHREADS
	char c;
#endif
	int64_t i;

	SRAND(time(NULL));
	for (i = 0; (numBlocks == 0 || i < numBlocks) && !flAborted; i++) {
#ifdef USE_PTHREADS
		MYASSERT(pthread_mutex_lock(&lock) == 0,
		    "pthread_mutex_lock failed");
		while (nblocks > 1 && !flAborted)
			MYASSERT(pthread_cond_wait(&less, &lock) == 0,
			    "pthread_cond_wait failed");
		MYASSERT(pthread_mutex_unlock(&lock) == 0,
		    "pthread_mutex_unlock");
#endif
		initblock(buf[i & 1], blockSize, type, i);
#ifdef USE_PTHREADS
		MYASSERT(pthread_mutex_lock(&lock) == 0,
		    "pthread_mutex_lock failed");
		nblocks++;
		MYASSERT(pthread_cond_signal(&more) == 0,
		    "pthread_cond_signal failed");
		MYASSERT(pthread_mutex_unlock(&lock) == 0,
		    "pthread_mutex_unlock");
#else
		MYASSERT(write(ctl1[1], &c, 1) == 1, "Write on pipe failed");
		MYASSERT(read(ctl2[0], &c, 1) == 1, "Read on pipe failed");
#endif
	}
	return NULL;
}

static void *
status(void *dummy)
{
	while (!flAborted && !flFinished) {
		statusLine(blocksWritten * blockSize / 1024, numBlocks * blockSize / 1024,
		    "KiB", "KiB/s");
		usleep(STATUS_UPDATE_TIME);
	}
	fputc('\n', stderr);
	return 0;
}

static void
cleanup(int sig)
{
	flAborted = 1;
}

static void
usage()
{
	fprintf(stderr,
		"fblckgen version " PACKAGE_VERSION ".\n"
		"Copyright Paul Ripke\n"
#ifdef USE_PTHREADS
		"Built to use pthreads.\n\n"
#else   
		"Built to use multiple processes.\n\n"
#endif  
		"Usage: fblckgen [-a | -r] [-q] [-v] [-b bytes] [-c count]\n\n"
		"  -a          Write blocks of a repeating ASCII "
		    "string (compresses well)\n"
		"  -r          Write blocks of binary 'random' data "
		    "(shouldn't compress)\n"
		"  -b bytes    Set write blocksize\n"
		"  -c count    Number of blocks to write "
		    "(zero for infinite)\n"
		"  -q          Quiet operation\n"
		"  -v          Display progress line\n\n"
		"Compiled defaults:\n"
		"    fblckgen -a -b 1s -c 1k\n\n"
		"Numeric arguments take an optional "
		    "letter multiplier:\n"
		"  s:        Sectors (x 512)\n"
		"  k:        kibi (x 1024 or 2^10)\n"
		"  m:        mebi (x 1048576 or 2^20)\n"
		"  g:        gibi (x 2^30)\n"
		"  t:        tebi (x 2^40)\n"
		"  p:        pebi (x 2^50)\n"
		"  e:        exbi (x 2^60)\n"
	);
}
